### PR TITLE
Create view combining flow counts with those of their replacement sensors

### PR DESCRIPTION
**Pull request by Nate-Wessel**
## What this pull request accomplishes:

- Some ecocounter sensors were replaced by other ecocounter sensors. This will string the counts together between the old and new sensors. 
- Adds view `ecocounter.combined_flows` which recursively combines ecocounter sensors. 

## Issue(s) this solves:

working toward resolving https://github.com/Toronto-Big-Data-Innovation-Team/monitoring/issues/35

## What, in particular, needs to reviewed:

-  column naming

## What needs to be done by a sysadmin after this PR is merged

NA

### CODE REVIEW

**Comment by gabrielwol** [comment]
Here's a test script to show the recursive functionality in action, including for recursion beyond level 2 (which we don't have any examples of currently). 
```sql
CREATE TABLE flows (flow_id integer, replaced_by_flow_id integer);
INSERT INTO flows 
VALUES
    (1, null),
    (2, 1),
    (3, 2),
    (4, 2),
    (5, null);

WITH RECURSIVE recursive_flows AS (
    --flows which have never been replaced
    SELECT flow_id, replaced_by_flow_id
    FROM flows
    WHERE replaced_by_flow_id IS NULL

    UNION

    --flows replaced by those in above recursive CTE.
    SELECT
        f.flow_id,
        COALESCE(rf.replaced_by_flow_id, rf.flow_id)
    FROM flows AS f
    JOIN recursive_flows AS rf ON rf.flow_id = f.replaced_by_flow_id
)

--all flows and their current replacement
SELECT
    flow_id,
    COALESCE(replaced_by_flow_id, flow_id) AS current_flow_id
FROM recursive_flows
ORDER BY flow_id;
```

**Comment by gabrielwol** [comment]
Here's a example usage, which combines data from flow_ids 104042942 (replaced by 353341333) and 353341333. Do the column names make sense @Nate-Wessel?
```sql
SELECT datetime_bin::date, SUM(volume)
FROM ecocounter.combined_flows
JOIN ecocounter.counts USING (flow_id)
WHERE current_flow_id = 353341333
GROUP BY 1
ORDER BY 1
```

**Comment by Nate-Wessel** [comment]
> Here's a example usage, which combines data from flow_ids 104042942 (replaced by 353341333) and 353341333. Do the column names make sense @Nate-Wessel?
> 
> ```sql
> SELECT datetime_bin::date, flow_id, SUM(volume)
> FROM ecocounter.combined_flows
> JOIN ecocounter.counts USING (flow_id)
> WHERE current_flow_id = 353341333
> GROUP BY 1, 2
> ORDER BY 1
> ```

That makes sense. And also, it looks like we found another anomalous range!

graph_visualiser-1721654548222

**Review by Nate-Wessel — COMMENTED** [review]
I approve. A couple tiny notes on the code; not important. 

One thought: should we remove the `replaced_by_site_id` field? I think we're right to implement this at the level of flows. 

One more thought: Should we pass the newly combined flows through the `flows` and `counts` views? I.e. hide the complexity of possible replacement from a user? I'd be happy to see this merged though, and that question can be answered in a separate issue later.

File: volumes/ecocounter/views/create-view-combined_flows.sql [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        flow_id,
+        COALESCE(replaced_by_flow_id, flow_id) AS current_flow_id
+    FROM recursive_flows
+    ORDER BY flow_id
+);
+
+ALTER VIEW ecocounter.combined_flows OWNER TO ecocounter_admins;
+GRANT ALL ON TABLE ecocounter.combined_flows TO ecocounter_admins;
+
+REVOKE ALL ON TABLE ecocounter.combined_flows FROM bdit_humans;
```
[the remark this answers is not in the document]
Comment (gabrielwol) [reply volumes/ecocounter/views/create-view-combined_flows.sql:31]: When you create a new object, postgres implements these permissions automatically:

```sql
ALTER TABLE IF EXISTS gwolofs.test OWNER to gwolofs;
GRANT ALL ON TABLE gwolofs.test TO gwolofs;

REVOKE ALL ON TABLE gwolofs.test FROM bdit_humans;
GRANT SELECT ON TABLE gwolofs.test TO bdit_humans;
```

**Comment by gabrielwol** [comment]
> One thought: should we remove the replaced_by_site_id field? I think we're right to implement this at the level of flows.

I'm ok to leave it, I think it still makes sense when reviewing sensor inventory even if we're not using it for this purpose. 

> One more thought: Should we pass the newly combined flows through the flows and counts views? I.e. hide the complexity of possible replacement from a user? I'd be happy to see this merged though, and that question can be answered in a separate issue later.

What do you think about adding `current_flow_id` to those views, so you can reference either that or `flow_id` (original)?

**Comment by Nate-Wessel** [comment]
> I'm ok to leave it, I think it still makes sense when reviewing sensor inventory even if we're not using it for this purpose.

Noted this a little more clearly in the docs.

**Comment by Nate-Wessel** [comment]
> What do you think about adding `current_flow_id` to those views, so you can reference either that or `flow_id` (original)?

I think I'd rather do it the other way around if anything. Like `flow_id` and `old_flow_id`. The user shouldn't need to understand why the `flow_id` may change to use the data correctly.

File: volumes/ecocounter/readme.md
Comment (Nate-Wessel) [inline volumes/ecocounter/readme.md:1]: Shuffled some fields around a bit here. Their order in the table seemed a bit random.

**Review by gabrielwol — APPROVED** [review]
Nate and I are in consensus on merging this one ✅. Any further views can be created in separate issues as necessary.

**Pull request merged**